server/src/main/java/com/epam/aidial/core/server/util/ApplicationTypeSchemaUtils.java

```diff
@@ -0,0 +1,212 @@
+    }
+
+    public static String getCustomApplicationEndpoint(Config config, Application application) {
+        try {
+            String schema = getCustomApplicationSchemaOrThrow(config, application);
+            JsonNode schemaNode = ProxyUtil.MAPPER.readTree(schema);
+            JsonNode endpointNode = schemaNode.get("dial:applicationTypeCompletionEndpoint");
+            if (endpointNode == null) {
+                throw new ApplicationTypeSchemaProcessingException("Custom application schema does not contain completion endpoint");
+            }
+            return endpointNode.asText();
+        } catch (JsonProcessingException | IllegalArgumentException e) {
+            throw new ApplicationTypeSchemaProcessingException("Failed to get custom application endpoint", e);
+        }
+    }
+
+    public static Application modifyEndpointForCustomApplication(Config config, Application application) {
+        String customEndpoint = getCustomApplicationEndpoint(config, application);
+        if (customEndpoint == null) {
+            return application;
+        }
+        Application copy = new Application(application);
+        copy.setEndpoint(customEndpoint);
+        return copy;
+    }
+
+    public static Application filterCustomClientProperties(Config config, Application application) {
+        String customApplicationSchema = getCustomApplicationSchemaOrThrow(config, application);
+        if (customApplicationSchema == null) {
+            return application;
+        }
+        Application copy = new Application(application);
+        Map<String, Object> appWithClientOptionsOnly = filterProperties(application.getCustomProperties(), customApplicationSchema, "client");
+        copy.setCustomProperties(appWithClientOptionsOnly);
+        return copy;
+    }
+
+    public static Application filterCustomClientPropertiesWhenNoWriteAccess(ProxyContext ctx, ResourceDescriptor resource, Application application) {
+        if (!ctx.getProxy().getAccessService().hasWriteAccess(resource, ctx)) {
+            application = filterCustomClientProperties(ctx.getConfig(), application);
+        }
+        return application;
+    }
+
+    public static void replaceCustomAppFiles(Application application, Map<String, String> replacementLinks) {
+        if (application.getCustomAppSchemaId() == null) {
+            return;
+        }
+        JsonNode customProperties = ProxyUtil.MAPPER.convertValue(application.getCustomProperties(), JsonNode.class);
+        replaceLinksInJsonNode(customProperties, replacementLinks, null, null);
+        Map<String, Object> customPropertiesMap = ProxyUtil.MAPPER.convertValue(customProperties, new TypeReference<>() {
+        });
+
+        application.setCustomProperties(customPropertiesMap);
+    }
+
+    @SuppressWarnings("unchecked")
+    public static List<ResourceDescriptor> getFiles(Config config, Application application, EncryptionService encryptionService, ResourceService resourceService) {
```

Comment:
Could we move methods like that (with dependencies to services) to a separate class like ApplicationTypeSchemaService?

Reply:
We can. It started with lol less code, but then we moved almost everything there) So, I suggest accepting that and then moving while refactoring.